File: clients/typescript/src/utils/bytes.ts
export function takeBytes(N: number, buf: Uint8Array): [Uint8Array, Uint8Array] {
    if (N > buf.length) {
        throw new Error(`takeBytes(${N}) exceeds buffer length ${buf.length}`);
    }
    return [buf.subarray(0, N), buf.subarray(N)];
}

export class Bytes {
    static empty() {
        return new Bytes(new Uint8Array())
    }

    constructor(private data: Uint8Array) { }

    get length() {
        return this.remaining().length
    }

    isEmpty() {
        return this.length == 0
    }

    nextByte() {
        let [[byte], ptr] = takeBytes(1, this.data);
        this.data = ptr;
        return byte
    }

    take(len: number) {
        let [bytes, ptr] = takeBytes(len, this.data);
        this.data = ptr;
        return bytes;
    }

    remaining() {
        return this.data;
    }
}

File: clients/dart/lib/utils/bytes.dart
import 'dart:typed_data';

import 'package:setu_client/utils/common.dart';

(Uint8List, Uint8List) takeBytes(int n, Uint8List buf) {
  assertExpr(
    n <= buf.length,
    error: RangeError.new,
    message: () => 'takeBytes($n) exceeds buffer length ${buf.length}',
  );
  return (Uint8List.sublistView(buf, 0, n), Uint8List.sublistView(buf, n));
}

class Bytes {
  factory Bytes.empty() {
    return Bytes(Uint8List(0));
  }

  Bytes(this._data);

  Uint8List _data;

  int get length => _data.length;
  bool get isEmpty => _data.isEmpty;
  Uint8List get remaining => _data;

  int nextByte() {
    final (bytes, ptr) = takeBytes(1, _data);
    _data = ptr;
    return bytes[0];
  }

  Uint8List take(int len) {
    final (bytes, ptr) = takeBytes(len, _data);
    _data = ptr;
    return bytes;
  }
}
